File: src/main/java/com/meetify/server/controller/UserRestController.java
package com.meetify.server.controller;

import com.meetify.server.model.Id;
import com.meetify.server.model.User;
import com.meetify.server.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.Arrays;
import java.util.HashSet;
import java.util.Set;

/**
 * com.meetify.server.controller
 * Created by kr3v on 02.10.2016.
 */

@RestController
@RequestMapping("/user")
public class UserRestController {
    private final UserRepository userRepository;

    @Autowired
    public UserRestController(UserRepository userRepository) {
        this.userRepository = userRepository;
    }
    
    private User get(Long id) {
        return userRepository.findById(new Id(id)).orElse(null);
    }

    @RequestMapping(method = RequestMethod.GET)
    public ResponseEntity<?> get(@RequestParam(value = "id") String id) {
        StringBuffer stringBuffer = new StringBuffer("");
        Arrays.stream(id.split(",")).forEach(s -> stringBuffer.append(get(Long.parseLong(s))));
        return new ResponseEntity<>(stringBuffer, null, HttpStatus.OK);
    }

    @RequestMapping(method = RequestMethod.POST)
    public ResponseEntity<?> post(@RequestParam(value = "id") String id,
                                  @RequestParam(value = "friends", defaultValue = "") String friends) {
        Set<Id> set = new HashSet<>();
        if (!friends.equals("")) {
            Arrays.stream(friends.replaceAll("[^,\\d]", "").split(",")).forEach(s -> set.add(new Id(Long.parseLong(s))));
        }
        return new ResponseEntity<>(userRepository.save(new User(new Id(Long.parseLong(id)), set)), null, HttpStatus.OK);
    }
}

File: src/main/java/com/meetify/server/controller/UserRestController.kt
package com.meetify.server.controller

import com.meetify.server.model.Id
import com.meetify.server.model.User
import com.meetify.server.repository.UserRepository
import com.meetify.server.utils.JsonUtils
import com.meetify.server.utils.JsonUtils.mapper
import com.meetify.server.utils.MultiSearchable
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.web.bind.annotation.*
import java.util.*

/**
 * This class represents controller over users. It holds mapping '/user'.
 * @author  Dmitry Baynak
 * @version 0.0.1
 * @since   0.0.1
 * @param   userRepository  users repository, which is provided by Spring & Hibernate.
 * @constructor             Autowired by Spring.
 */
@RestController @RequestMapping("/user")
class UserRestController @Autowired constructor(private val userRepository: UserRepository) : MultiSearchable<User>() {
    /**
     * Returns a list with information about users.
     * Unknown ids in [idsJson] list are ignored.
     * @param   idsJson json representation of collection of the ids of requested users.
     * @return          collection with users info.
     */
    @ResponseBody @RequestMapping(method = arrayOf(RequestMethod.GET))
    fun get(@RequestParam(name = "ids") idsJson: String): ArrayList<User> {
        return get(userRepository, JsonUtils.getList(idsJson))
    }

    /**
     * Returns a list with information about friends of selected user.
     * If some friends are unknown, they are ignored.
     * @param   idJson  json representation of id of user, which friends should be returned.
     * @return          list with friends info.
     */
    @ResponseBody @RequestMapping("/friends", method = arrayOf(RequestMethod.GET))
    fun friends(@RequestParam(name = "id") idJson: String): ArrayList<User> {
        val id = mapper.readValue(idJson, Id::class.java)
        var friends = ArrayList<User>()
        userRepository.findById(id).ifPresent { friends = get(userRepository, it.friends) }
        return friends
    }

    /**
     * Put method, which is used to upgrade some info about user.
     * Mainly, should be used to send to server info about current location of this [user].
     * If [user] in body of request contains non-empty list of friends id, it also updated.
     * If some info about places (not location) in body is given, it's ignored.
     * @param   user  user, which should be updated with given information.
     */
    @ResponseBody @RequestMapping("/update", method = arrayOf(RequestMethod.PUT))
    fun update(@RequestBody user: User) {
        userRepository.findById(user.id).ifPresent {
            if (user.friends.size > 0) {
                it.friends = user.friends
            }
            it.location = user.location
            userRepository.save(it)
        }
    }

    /**
     * Method, that should be used only to register new users.
     * If [user] with the same id was found, given in request body user is ignored, and saved [user] is returned.
     * Otherwise, new [user] will be saved with given information.
     * If some info about places (not location) is given, it's ignored.
     * @param   user  user, which should be created.
     */
    @ResponseBody @RequestMapping(method = arrayOf(RequestMethod.POST))
    fun post(@RequestBody user: User): User {
        return userRepository.save(userRepository.findById(user.id).orElse(User(user.id, user.location, user.friends)))
    }
}